Effects History View for Account Page

robertDurst opened the pull request:
Added a tab to the account page linking to a view with a log of the user's effects history. The log is organized into a table in reverse chronological order and able to be filtered by effect type ('account', 'signer', 'trade', 'trustline'). It displays the type of effect, the action (ie for account created, it would be created), a description of the effect, and the date (with ledger sequence number for further reference). Here is a gif:

![tabledemo](https://user-images.githubusercontent.com/16689974/33236715-0e07e506-d211-11e7-9064-261b5d6d9d0e.gif)

I did my best to maintain the styling of the current codebase and include only necessary code to create the additional view. 

***Note: Referencing #80. ***

irisli commented:
While gifs are cool for showing off, they're terrible at describing what was done since I have to watch a 20 second thing to just see a small 1 second glimpse of something. I suggest using many static images when showing something. Only use gifs when you want to show something off but have something to hide (kinda shady, but thats what gifs are great for)

irisli commented on src/components/Session/AssetCardEffect.jsx:
```diff
@@ -0,0 +1,72 @@
+/*
```
This file is unecessary. Making a clone of AssetCard2 complicates the code for almost no benefits

irisli commented on src/components/Session.jsx:
```diff
@@ -75,6 +76,14 @@ class Session extends React.Component {
             <Send d={d}></Send>
           </div>
         </div>
+      } else if (part1 === 'transactionHistory') {
```
:+1: good

irisli commented on src/components/Session/DateCard.jsx:
```diff
@@ -0,0 +1,26 @@
+/*
```
Breaking things up into modules is great when things are used in many places. But when making something for the first time, just keep it simple in just one file.

Having so many types of cards is confusing especially since it's not actually a UI element.

Additionally, loading shouldn't be done by individual components. Instead, use https://github.com/irisli/stellarterm/blob/master/src/components/Loading.jsx

irisli commented on src/components/Session/EffectCard.jsx:
```diff
@@ -0,0 +1,151 @@
+/*
```
Here again, the effect card is un-necessary. Better to name it as a "TxHistoryRow". Card implies that it is a "card" like thing with many different uses.

irisli commented on src/components/Session/HistoryTable.jsx:
```diff
@@ -0,0 +1,77 @@
+  render() {
+    let account = this.props.d.session.account;
+
+    // From MagicSpoon.Account
```
State should be in the "Driver" somewhere: https://github.com/irisli/stellarterm/blob/master/src/lib/Driver.js

irisli commented on src/lib/Format.js:
```diff
@@ -37,3 +37,36 @@ exports.niceNumDecimals = function niceNumDecimals(input) {
   }
   return 7;
 };
+
```
Relative times are good for chats and social media where relative time is important but exact time is irrelevant. For financial applications, relative time is unimportant but exact time is important.

Lets keep things simple and just show the exact time relative to the browser's timezone

irisli commented on src/lib/MagicSpoon.js:
```diff
@@ -10,6 +10,32 @@ import directory from '../directory';
   async Account(Server, keypair, onUpdate) {
     let sdkAccount = await Server.loadAccount(keypair.publicKey())
+
+    let effectsHistory = await Server.effects().forAccount((keypair.publicKey())).limit(200).order('desc').call();
```
This seems quite wasteful to make up to 201 requests per account. Perhaps we can just load tx details when the user hits the transaction history page and then cache the results.

irisli commented on src/styles/_HistoryTable.scss:
```diff
@@ -0,0 +1,70 @@
+// This file contains all the styles for the Effects History Table Component: HistoryTable.jsx
```
Good job with following the css conventions :+1:

irisli commented:
Overall:

- Reduce code duplication
- Absolute timestamps
- More efficient fetching of effects

robertDurst commented:
My apologies!! Appearing shady was not my intention. I figured that a video/gif would be cool and act as proof that the actions described worked simultaneously and weren't hardcoded. Below are some statics highlighting the features of the table. I'll follow up with the feedback you've provided above in the next few days. Thanks!

## Overall View Screenshot
Here you can see the entirety of the effects history log. The main body is scrollable, and the header is static. The checkboxes allow you to filter between the different types. The main effect type category is bolded on the left for easy scalability and the sub type is to its right. After that you have a description of the effect along with data and ledger sequence number beneath it. Most people unfamiliar with Stellar won't necessarily understand the significance of ledger sequence numbers, so that is why I decided for it to go beneath the date and in smaller font.
<img width="1440" alt="screen shot 2017-11-28 at 11 36 19 am" src="https://user-images.githubusercontent.com/16689974/33340425-eb5e82ea-d430-11e7-9c82-3a785985c830.png">

## Addition of tab in the account links heading
I decided it would be logical to include a link to the history in the account heading. At first I thought maybe this table should be with the account balance, however some effects have nothing to do with balance, so I figured the history table should be entirely seperate.
<img width="1404" alt="screen shot 2017-11-28 at 11 49 56 am" src="https://user-images.githubusercontent.com/16689974/33341019-78cfeb9a-d432-11e7-85ec-2e7c9e88c748.png">

## Trade History
Here are some examples of trade history. Trade history utilizes the already implemented asset card and shows a simple visual of what currency was exchanged and for what amount.
<img width="984" alt="screen shot 2017-11-28 at 11 36 47 am" src="https://user-images.githubusercontent.com/16689974/33340248-77ddfcba-d430-11e7-9eec-e34e08ea48b1.png">

## Account History
Account history has many different sub effects. These sub effects include threshold changes, flag changes, transactions out, transactions in, and home domain changes. These are all represented in the screenshot below.
<img width="983" alt="screen shot 2017-11-28 at 11 38 31 am" src="https://user-images.githubusercontent.com/16689974/33340499-15430004-d431-11e7-99e6-856764528720.png">


## Trustline History
Here are the main trustline effects. Again, I utilize the asset card for a nice visual.
<img width="983" alt="screen shot 2017-11-28 at 11 38 12 am" src="https://user-images.githubusercontent.com/16689974/33340486-0cde3a5a-d431-11e7-9664-dd7b6278ecba.png">


## Signer History
I decided that for signer effects, the most useful information would be the key (pk) being used to update and the weight of the key.
<img width="985" alt="screen shot 2017-11-28 at 11 38 22 am" src="https://user-images.githubusercontent.com/16689974/33340480-08e23546-d431-11e7-895b-41f0e825eeaf.png">


## Filtering of History Log
Here is an example of filtering. As you can see, the trade and account effects are checked and those are the events that are displayed.
<img width="983" alt="screen shot 2017-11-28 at 11 38 51 am" src="https://user-images.githubusercontent.com/16689974/33340469-01eba272-d431-11e7-9767-69193f447749.png">

irisli commented:
- "Ledger sequence #12345" could be reduced to "Ledger #12345". Since sequence already implies number, the word sequence is redundant and we can save space
- I think the UI could be improved to be clearer but it's mergeable at this point. Perhaps I'll grind on how I could improve the UI. Perhaps there could even be multiple tables! But for now, lets just go with this current 1.0 UI. I now realize that the RippleTrade UI isn't as great as I imagined it to be. Thank you for helping me test this!

Good job on the screenshots though!

robertDurst commented:
Iris, I've gone through most of the changes you requested. However, still unsure about loading the effects. One way may be to load it when they first login. That way, it is cached and loaded by the time the user navigates to the history page. Then, I can have a listener that dynamically adds effects as they come in. That is sort of what I had originally.

Another idea I had was to load X amount at a time and display these page by page. This method is used by StellarChain.io and steexp.com and many other websites with tables. Loading 10 is pretty fast and once they are loaded, they can be cached.

Unless I am missing something, it appears the only way to get a substantial amount of information about any one effect is to do followup api calls for the individual effect. Thus, for 10 effects, I have to do 1 + 10 * 2 = 21 api calls. If there is a way around this, then loading would be quite trivial. 

Thanks!

robertDurst commented:
Smoothed out most of the errors above. Still working on an efficient and quick way of loading the table data. Right now I am thinking about putting it in a table that displays 8 results (looks good with current formatting) per page. Probably won't fix the remaining bugs until tomorrow evening and then I'll update the PR. Here is a screenshot of the current table design. Let me know if this is sort of what you were visualizing and if you think it fits the current design scheme.

### Table with Pages
<img width="1439" alt="screen shot 2017-11-30 at 2 14 17 pm" src="https://user-images.githubusercontent.com/16689974/33457993-54e58c3c-d5d9-11e7-98e0-5dfff5a0831f.png">

irisli commented:
Oh, oops. I mean it's good to load 200 transactions all at once. But I think it should only fetch details once we actually open the history page. I think a big long list is better than pagination because then it's searchable. But the pagination is good if we only show it for big transaction histories.

Additionally, can we show `hour:minute TZ` on the timestamps? I think that'd be useful

irisli commented:
![image](https://user-images.githubusercontent.com/5728307/33528397-935797ac-d814-11e7-877f-6e4e28e32f85.png)
This is amazing! Good job!

robertDurst commented:
Thanks, but still a few changes!! One moment, forgot that this pushes straight to the PR

irisli commented:
No problem. It's good to see work in progress

robertDurst commented:
All ready for review again! Here are the fixes in version 2:

## Time added
<img width="974" alt="screen shot 2017-12-03 at 12 21 24 pm" src="https://user-images.githubusercontent.com/16689974/33529346-8466db9e-d824-11e7-8b2f-ba2397749646.png">

Now we have absolute date and time. I agree this looks better and makes more sense for a financial log.

## AssetEffectCard Component merged with AssetCard2
Here I just took the meat of the Effect Card and put it in AssetCard2. This is some 6-8 lines of code. Then I added a check at the end for a isEffect property. If true, this loads the Effect Card. This works because the two cards have the same structure, but different styling.

## EffectCard renamed HistoryTableRow
Makes more sense.

## More Efficient Table Loading
First, I moved most of the table loading out of the account loading/signin function. Here is how the table now loads:

1) User signs in and a "skeleton" of the last 200 effects are loaded
2) User navigates to the History view and the most recent 7 effects are loaded and displayed. I chose 7 because that is how many are displayed on table at one time.
3) Right after the first 7 are loaded, the rest are loaded and added to the table in batches.

This significantly improved the table loading experience. Instead of waiting 3-5 seconds for everything to load, the user waits 0.3-1 seconds for the first "page" to load and then they can begin interacting with the table. Usually it is only another 1-3 seconds or the rest of the table to load. 

4) Once the table is loaded, it begins "listening" for new effects. These effects are added to the table.

Since the loading process saves the effect history, and since new effects are saved as well, the user can instantaneously filter through the history log, and will never have to wait again for the loading during this session. Also, it would now be easy to add a search bar in future versions.

## Known Issues
* Firefox does not do the overflow-y scrolling. Thus instead of a max 500 height Y scroll, it is just one long table.

* A warning for setState of an unmounted component when changing a trustline and then navigating back to the history view to watch the effect history update.

* I wasn't paying attention and pushed my Secret Key in the last commit. Nevertheless, that account has 75 lumens and was my testing account. If someone finds it, buy yourself something nice :)

irisli commented:
You could always go back in time to change the history. Though you might want to use the Stellar Laboratory to perform an account merge.

robertDurst commented:
It's all good, that was a throw away account.

robertDurst commented:
Let me know if there is anything else I should change or anything to redesign. Thanks!

robertDurst commented:
Well, I cut a couple hundred lines of code and then broke something. I'll fix this tomorrow.

robertDurst commented:
@irisli so the test that fails is the loading of the trade history on https://stellarterm.com/#exchange/.
This page won't even load on the live website for me; the website had a 404 error. I tested this on multiple devices and got the same result. 

So, I think my code may be fine after all.

irisli commented:
Yeah, the new horizon changes broke StellarTerm. Don't worry about the tests

irisli commented:
Okay, I scanned it and it all looks fine.

I'm going to manually change the code so that we don't use bluebird. It's doable without bluebird I believe.

irisli commented:
Hm, the feature is still glitchy and doesn't load properly. I'm going to have to fix it

robertDurst commented:
Ok, I'll check it out!

robertDurst commented:
What sort of glitchiness?

irisli commented:
Hey, I am halfway through rewriting it to fit into the existing code. I'll commit it and then you can work from my stopping point

irisli commented:
I think for the first version, we won't do streaming

robertDurst commented:
Ok sounds good. Streaming was an extra layer of difficulty

irisli commented:
Ok. The stuff is in the `history` tab. The stuff now loads properly and progressively.

Stuff I changed:
- Changed date ordering
- Removed EffectCard custom styles
- 

The whole layout of history row needs to be redone. The current way it's laid out is too messy and is just a blob of information.

I think it's fine if a row takes up more vertical space. It should be more clear to understand. Perhaps we don't even have assetcards and it'll only show up when the user overs over the asset name.

robertDurst commented:
Ok sounds good, I'll work on simplifying it. Would you say its simply too much information or just a poor organization of the information presented?

Also, the filtering is now broke. I'll fix that too.

robertDurst commented:
Here is a mockup for a simpler, more minimal chart. Wanted to get a better idea of what you are thinking before spending time coding it.

<img width="841" alt="screen shot 2018-01-02 at 10 53 09 am" src="https://user-images.githubusercontent.com/16689974/34495747-5d3b324c-efab-11e7-879a-71c4ff9ded02.png">

robertDurst commented:
Filter fixed, so functionality is there. Let me know your thoughts on the design.

irisli commented:
Yeah. Perhaps the `Received`, `Send`, `Traded` stuff can all be on its own line.

For the assets, call them `XLM-native` and `JPY-vcbear.net` and make them underlined. Then on hover, show assetcards below them. (I can implement that part for you)

robertDurst commented:
Ok sounds good. Here is an updated mockup:

<img width="815" alt="screen shot 2018-01-02 at 1 00 32 pm" src="https://user-images.githubusercontent.com/16689974/34499863-060b1642-efbd-11e7-9487-2d1eb8217f9c.png">

If you think that looks good, I can code it up so that the hover can be added.

irisli commented:
Yup. For the hover, here are some suggestions

```css
/* PSEUDOCODE. Take this and change it */

.HistoryView__asset {
  text-decoration: dotted;
  position: relative;
}
.HistoryView__asset__card {
  display: none;
}

.HistoryView__asset:hover .HistoryView__asset__card {
  display: block;
  position: absolute;
  bottom: -50px;
}
```

robertDurst commented:
New PR #116
